viewing_party/party.py

```diff
@@ -1,23 +1,212 @@
+
+    for movie in user_data["watched"]:
+        if movie["title"] in set_user_unique_titles:
+            user_unique_watched_movies.append(movie)
+            
+    return user_unique_watched_movies
+
+
+def get_friends_unique_watched(user_data: dict) -> list: 
+    user_watched_movies = user_wateched_movies(user_data)
+    friends_watched_movies = friends_wateched_movies(user_data)
+    unique_recomended_movies = []
+     
+    set_friends_unique_movie_titles = friends_watched_movies.difference(user_watched_movies) #  Set of movies only watched by friends but not user
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] in set_friends_unique_movie_titles:
+                if movie["title"] not in unique_recomended_movies:
+                    unique_recomended_movies.append(movie)
+    
+    unique_recomended_movies = remove_duplicate_values(unique_recomended_movies)
+
+    return unique_recomended_movies
+          
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data: dict) -> dict:
+    user_movies_not_watched = get_friends_unique_watched(user_data) 
+    user_subscriptions = user_data["subscriptions"]
+    recommended_movies = []
+
+    for movie in user_movies_not_watched:
+        if movie["host"] in user_subscriptions:
+            recommended_movies.append(movie)
+            
+    return recommended_movies
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def is_empy(user_data: dict) -> list:
+    is_empty = list()  
+
+    for friend in user_data["friends"]:
+        if friend["watched"]:
+            is_empty.append(False)
+        is_empty.append(True)
+    
+    return is_empty
+
+
+def user_most_frequent_watched_genre(user_data: dict) -> str:
+    list_user_most_watched_genres = list()
+
+    if len(user_data["watched"]) == 0:
+        return list_user_most_watched_genres
+
+    for movie in user_data["watched"]:
+        list_user_most_watched_genres.append(movie["genre"])
+    
+    most_watched_genre = max(set(list_user_most_watched_genres), key=list_user_most_watched_genres.count)
+    
+    return most_watched_genre
+
+
+def get_new_rec_by_genre(user_data: dict) -> list:
+    friends = user_data["friends"] 
+    user_most_watched_genre = user_most_frequent_watched_genre(user_data)
+    recomended_movies = list()
+    is_empty = is_empy(user_data)   
+
+    if len(user_data["watched"]) == 0:
+        return recomended_movies
+        
+    if all(is_empty):
+        return recomended_movies
+
+    friends_movie_watched = get_friends_unique_watched(user_data)
+    for movie in friends_movie_watched:
+        if movie["genre"] == user_most_watched_genre:
+            recomended_movies.append(movie)
+
+    return recomended_movies

```

Comment:
👍 Great work considering if a friend had an empty watched list. I would recommend moving that logic over to `get_friends_unique_watched` just to maintain the single responsibility principle for this function to _only_ get recommendations by genre rather than also checking the integrity of the friends watched list. 

Additionally, if only one of the friends had an empty watched list but the rest of the friends had movies, should we return an empty list? (Your conditional `if all(is_empty)` assumes that.

Reply:
Yes, If all friends had an empty watched list, it'd return and empty list.